packages/accounts-controller/src/AccountsController.ts

```diff
@@ -477,50 +477,77 @@ export class AccountsController extends BaseController<
    * @returns A Promise that resolves when the accounts have been updated.
    */
   async updateAccounts(): Promise<void> {
-    const snapAccounts = await this.#listSnapAccounts();
-    const normalAccounts = await this.#listNormalAccounts();
-
-    // keyring type map.
-    const keyringTypes = new Map<string, number>();
-    const previousAccounts = this.state.internalAccounts.accounts;
-
-    const accounts: Record<string, InternalAccount> = [
-      ...normalAccounts,
-      ...snapAccounts,
-    ].reduce((internalAccountMap, internalAccount) => {
-      const keyringTypeName = keyringTypeToName(
-        internalAccount.metadata.keyring.type,
-      );
-      const keyringAccountIndex = keyringTypes.get(keyringTypeName) ?? 0;
-      if (keyringAccountIndex) {
-        keyringTypes.set(keyringTypeName, keyringAccountIndex + 1);
+    // We are re-creating the list of internal accounts based on the accounts from
+    // each keyrings, following those rules:
+    //
+    // - If the account already exists on this controller state, we only update
+    //   metadata. This allows to automatically "migrate" new metadata for existing
+    //   accounts.
+    //
+    // - Else, we create the associated internal account for that new account. This
+    //   can happen when the Snap keyring adds a new account. The Snap keyring will
+    //   create a new account and will automatically "persists" by calling this
+    //   function explicitly to create its associated internal account.
+    //   FIXME: This logic is kinda tricky and could be replaced with the new
+    //          incoming notification system that the Snap keyring will use.
+    const snapAccounts = await this.#listAccountsFromSnapKeyring();
+    const normalAccounts = await this.#listAccountsFromOtherKeyrings();
```

Comment:
Maybe it's worth to rename `normalAccounts` to something different, it doesn't sound right. Maybe `nativeAccounts` or `nonSnapAccounts`.

Reply:
I do agree yes. I'll rename this.

Reply:
As discussed internally. I'll do this change in another PR to avoid having too many "unrelated" changes on this one, see:
- https://github.com/MetaMask/core/pull/5162